### ISSUE TITLE
Proposal: Add Python support to WinUI 3

### ISSUE THREAD

**Issue by freakboy3742** [microsoft/microsoft-ui-xaml#3364]
# Proposal: Add Python support to WinUI3

## Summary

WinUI 3 should include an official Python API, allowing end-users to build native Windows GUI apps in Python.

## Rationale

 * Python is one of the most popular programming languages in the world - by some measures, *more* popular than [JavaScript, C# and C++](https://www.tiobe.com/tiobe-index/).
 * "Developing a native GUI app" is one of the most fundamental use cases for *any* programming language. However, there are no Microsoft supported paths at present for build Windows GUIs using Python. [Pythonnet](https://github.com/pythonnet/pythonnet) is a third party tool that *can* be used, but it doesn't produce UWP-compatible apps.
 * Python is widely used in education as an introductory language for new developers. Giving students the ability to build GUIs is an essential part of a modern developer experience.
 * Python is widely used in scientific circles to support research. These researchers currently use Jupyter to provide visualisations of there data analysis; however, there Jupyter cannot be used to develop standalone applications for visualization or data gathering. A native GUI API would enable scientists to develop these apps.
 * Microsoft already has an established interest in Python as a language. Microsoft has a dedicated Python language team (employing multiple members of the Python core team); VSCode has established Python language support; and Microsoft sponsors multiple major Python conferences.

## Important Notes

This has been proposed in direct response to [this tweet from @WindowsDocs](https://twitter.com/WindowsDocs/status/1311373139708768256), and [this follow up from @zooba](https://twitter.com/zooba/status/1311948449114357761). It may be related to #1421.

**Comment by Felix-Dev** [comment]
Probably should also link to https://github.com/microsoft/xlang/tree/master/src/package/pywinrt/projection which could see the necessary work required for WinUI support (as in: [Xaml support](https://github.com/microsoft/xlang/tree/master/src/package/pywinrt/projection#known-issues)).

**Comment by MarcSkovMadsen** [comment]
This could be really awesome for data apps. Please think about how to integrate the wide range of popular Dataviz tools like Matplotlib, plotly, bokeh, holoviews, datashader etz.

And look to the popular web frameworks bokeh, Dash, panel, Streamlit and voila for inspiration on how domain experts, scientists and data scientist develop data apps.

For example check out my sites awesome-panel.org and awesome-panel.org.

**Comment by pvcraven** [comment]
Make sure we can get an OpenGL context, so libraries like Arcade can do high-performance graphics.

**Comment by laike9m** [comment]
+1

**Comment by RushNMob** [comment]
Here to show support for this.

**Comment by miguelp1986** [comment]
Please add Python support to WinUI 3!

**Comment by AzizAlqasem** [comment]
In my opinion, If Microsoft wants to substantially increase the number of apps that support windows, then one of the best approaches is to invest in the Python community. They are huge in number, enthusiasm, ambition, and growth. A large portion of them are young, and young developers are the future! Your app store will be full of python based apps :)

As a Python developer, if Windows supports Python, then that will make it my first choice as a development OS.

Best,

**Comment by leadersheir** [comment]
Strongly in favor 🤚

**Comment by kilrogg** [comment]
Absolutely Support this.

**Comment by roodie** [comment]
I support this idea!

**Comment by BertRaeymaekers** [comment]
Yes please.

**Comment by holg** [comment]
Yes i'd very much appreciate.

**Comment by 70Shubham07** [comment]
Totally support this!

**Comment by Abdur-rahmaanJ** [comment]
Support from Mauritius

**Comment by Twilight0** [comment]
I also support this suggestion!

**Comment by SFM61319** [comment]
Hey iirc there *is* a python library for this.  It's called [`pythonnet`](https://github.com/pythonnet/pythonnet), a python library for the .NET framework.

**Comment by hasii2011** [comment]
Please add Python support to WinUI 3

**Comment by andycrofford** [comment]
I support adding Python support to the WinUI3 library.

**Comment by tfillmor924** [comment]
I support this!!!

**Comment by mkturner** [comment]
I'm interested in Python Support for WinUI3.

With recent additions like WSL and native containers Windows has started talking my language. I'd honestly strongly consider Windows as a daily driver if Python was picked up officially.

**Comment by rlitchfield-Edcara** [comment]
This would be amazing addition to the platform and I'm 100% for it.

**Comment by cerickson** [comment]
This would be wonderful!

**Comment by frisbeegrammer** [comment]
Yes please !

**Comment by rednafi** [comment]
This is gonna be awesome!

**Comment by StephenLPeters** [comment]
@benJkuhn and @kennykerr FYI! :)

**Comment by iCarlosCode** [comment]
We shall not lose this opportunity! This can change the world for the better!

**Comment by awildtechno** [comment]
This is needed! Windows is falling behind in the apps race because it doesn't have a straightforward solution for beginners to try their hand at making native apps.

**Comment by AzizAlqasem** [comment]
I am very excited to see Windows (WinUI3) supports Python, and I hope Microsoft approve that as soon as possible!

**Comment by aeroaks** [comment]
Have developed GUIs with Python on Windows platform for few years. Had to use win32 and some other libraries to use specific functionalities. This support would have helped a lot.

**Comment by aeroaks** [comment]
Have developed GUIs with Python on Windows platform for few years. Had to use win32 and some other libraries to use specific functionalities. This support would have helped a lot.

**Comment by arcanosam** [comment]
Would be awesome indeed 👐✌😀

**Comment by StephenLPeters** [comment]
@ryandemopoulos and @pag3 fyi as well

**Comment by ArkNieckarz** [comment]
I have legacy win32 apps that are waiting to be redone in Python and WinUI 3 will be needed.
Yes very interested in Python Support for WinUI3.

**Comment by alik472** [comment]
Yes Please!

**Comment by ht-thomas** [comment]
I definitely cannot wait for this feature to role out...

**Comment by NyntoFive** [comment]
YesYesYes

**Comment by vsergeyev** [comment]
Yes please! Python applications for Machine Learning may greatly benefit from this.

**Comment by gruntfutuk** [comment]
That would be awesome.

**Comment by vednig** [comment]
👍

**Comment by paaKways** [comment]
I support this!

**Comment by samueljohn** [comment]
would be cool, indeed!!

**Comment by anhelus** [comment]
It would be great!

**Comment by Roman-username** [comment]
yes!

**Comment by Jaiganeshkumaran** [comment]
This should be done. Python has become one of the most popular programming languages and today there isn't a good way t0o build modern Windows apps using Python. Windows Runtime was designed with cross-language support with mind and not to just access Windows APIs in OOP way in C#.

**Comment by hussain5416** [comment]
Here to show the support +1

**Comment by 42837531** [comment]
> Here to show the support +1

There is a like button.

**Comment by iCarlosCode** [comment]
> > Here to show the support +1
> 
> There is a like button that is more suited for that purpose than the comment feature.

Just in case, let's comment and click the like button. Hahaha

**Comment by wiryonolau** [comment]
Either this or invaded by electron

**Comment by sumeshir26** [comment]
Please, would be a great alternative to Tkinter which looks very outdated.

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
[http://github.com/microsoft/microsoft-ui-xaml/issues/3364#issuecomment-877989796](url)
For Windows, Tkinter.ttk is a good alternative until support for Python is added to WinUI 3.

**Comment by aarmn** [comment]
Support for this proposal and beeware project

**Comment by iamkartiknayak** [comment]
Please make it happen!!

**Comment by pbk0** [comment]
I am from India ... and can even forget our animosity towards Pakistan if you support this ..... haha

**Comment by alireza138812** [comment]
So please do it

**Comment by benoitlathiere** [comment]
Please add Python support to WinUI 3. Wxpython is not beautiful..
👍🤞

**Comment by hussain5416** [comment]
This issue is older than 1 year, still there's no response from WinUI team. I do not know what they are planning, but it is always a good idea to provide some details on such a popular proposal.

**Comment by hussain5416** [comment]
> As of 28 Jul 2021, the pip package winui was reserved by Microsoft Corporation: https://pypi.org/project/winui/ Hopefully this means that they are working towards a package?

Yes indeed, but there's no update from WinUI team on this issue. This is not fair for such old and popular issue.

**Comment by Fluent-Aero** [comment]
貌似开发一个能用Python调用C++接口的中间函数就可以了？

一定不止这么简单吧

**Comment by miketech** [comment]
+1

**Comment by miketech** [comment]
> Here from 2022.
> 
> As of 28 Jul 2021, the pip package winui was reserved by Microsoft Corporation: https://pypi.org/project/winui/ Hopefully this means that they are working towards a package?

Hope so 😁👍

**Comment by Aldrich-Code** [comment]
> 貌似开发一个能用Python调用C++接口的中间函数就可以了？
> 
> 一定不止这么简单吧

Python很多库就是这么干的

**Comment by dylech30th** [comment]
WinUI team needs more developers, from what I can see it's already been hard for them to triage and solve the huge amount of issues. Since the current WinUI 3 itself is really half-baked, I think the priority target of the team would be implementing features and handling bug&performance issues, I suspect that a proposal like foreign language support must have a relatively low priority until they make the WinUI 3 full-featured. Having said that, if anyone wants to get a quick reply, I suggest you post a comment in #6726 , it would be much more noticeable for them

**Comment by pjmlp** [comment]
They already can't manage just with C# and C++, let alone adding more languages into the mix.

**Comment by saint4eva** [comment]
> WinUI team needs more developers, from what I can see it's already been hard for them to triage and solve the huge amount of issues. Since the current WinUI 3 itself is really half-baked, I think the priority target of the team would be implementing features and handling bug&performance issues, I suspect that a proposal like foreign language support must have a relatively low priority until they make the WinUI 3 full-featured. Having said that, if anyone wants to get a quick reply, I suggest you post a comment in #6726 , it would be much more noticeable for them

That is what you get when a team decides to bet the feature of their on a less productive stack - C

> Here from 2022.
> 
> As of 28 Jul 2021, the pip package winui was reserved by Microsoft Corporation: https://pypi.org/project/winui/ Hopefully this means that they are working towards a package?

WinUI itself is far from being feature complete. Do you think they would ever touch that python package again? CS/WinRT, CS/ Win32 and C++/WinRT have many issues that need to be resolved.

**Comment by wherewhere** [comment]
After J#, C#, F# and B# (VB.NET), it will comes out P#?

**Comment by MosGeo** [comment]
Just thinking outloud (not an expert): would something like using pybind11 [https://github.com/pybind/pybind11](https://github.com/pybind/pybind11) work for this case?

**Comment by aarmn** [comment]
> After J#, C#, F# and B# (VB.NET), it will comes out P#? 😜

Ironpython was enough of a fail for them, I guess

**Comment by contextfree** [comment]
> After J#, C#, F# and B# (VB.NET), it will comes out P#? 😜

WinUI is WinRT-based, not .NET-based, so no. 🤖

**Comment by wherewhere** [comment]
> WinUI is WinRT-based, not .NET-based, so no. 🤖

P++ 😙

**Comment by lmjaedentai** [comment]
Fully agree with that
We love python

**Comment by SpazElectro** [comment]
Agreed.

**Comment by aarmn** [comment]
years and nothing but random notifications from this

**Comment by P-PPPP** [comment]
Agree

**Comment by monkeycc** [comment]
2023 
Agree

**Comment by Ianzb** [comment]
agree

**Comment by tiagorangel2011** [comment]
Yes please

**Comment by JasonMo1** [comment]
please

**Comment by JasonMo1** [comment]
Friends, I have found this, it can use some fluent widgets in PyQt or Pyside. https://github.com/zhiyiYo/PyQt-Fluent-Widgets

**Comment by tiagorangel2011** [comment]
Looks great, thanks for the link!

**Comment by yizigezi** [comment]
感觉需要Microsoft先出一个python与dotnet互通的库。目前仅有第三方实现，也许需要先做出这个再考虑WinUI3 API

**Comment by SpamixOfficial** [comment]
As much as I love linux and hate microsoft, PLEASE MICROSOFT! As a developer I prioritize user experience and I want all users to feel like they are at home when they are using my applications, and thats why I want to use winui for my application. Sadly though, it is not YET implemented, do I hope they do it soon.

**Comment by sayyid5416** [comment]
It's been **3 LONG YEARS**, and not a single reply from Microsoft on this thread. Well, that's very disappointing.

**Comment by frinkifail** [comment]
add it already :(

**Comment by sotanakamura** [comment]
I found that we can use WinUI 3 controls with Python. Now PyWinRT and I are working some problems to support it.

**Comment by SpamixOfficial** [comment]
> I found that we can use WinUI 3 controls with Python. Now PyWinRT and I are working some problems to support it.

Amazing! Please mention me when you've got a solution!

**Comment by SpamixOfficial** [comment]
> It's been **3 LONG YEARS**, and not a single reply from Microsoft on this thread. Well, that's very disappointing.

Tbh they are Microsoft, what do we expect?

**Comment by sotanakamura** [comment]
I found pywinrt cannot treat WinUI well because it is constructed on top of the abstraction of cppwinrt. I decided to create a new winmd parser to generate a pure python projection without C++. It may take some time.

**Comment by sotanakamura** [comment]
## Progress report

### Background

There are several python language projection for Windows Runtime. [PyWinRT](https://github.com/pywinrt/pywinrt) is one of the python projection for Windows Runtime API. It is constructed on top of the C++/WinRT and it has a technical problem that it cannot treat directly the internal mechanism of Windows Runtime due to the lack of knowledge of Windows Runtime. I decided to create a new pure python projection, but it takes a lot of time to make it ready to use.

### Another work

 I found another project [py-win32more](https://github.com/ynkdir/py-win32more). It is a pure python library providing a python language projection for Win32 APIs based on [win32metadata](https://github.com/microsoft/win32metadata). It can treat the internal mechanism of Windows Runtime completely. We will add the support of WinUI 3 soon and I think you can use WinUI 3 easily in Python in the future.

### Sample

This is improved in the future especially for constructors.

```py
from win32more.xaml import XamlApplication
from win32more.Microsoft.UI.Xaml import Window
from win32more.Microsoft.UI.Xaml.Controls import ColorPicker

class App(XamlApplication):
    def OnLaunched(self, args):
        win = Window.CreateInstance(None, None)
        clr = ColorPicker.CreateInstance(None, None)
        win.Content = clr
        win.Activate()

XamlApplication.Start(App)
```

**Comment by saint4eva** [comment]
@sotanakamura You guys are doing amazing work. Can we write the UI in code - in python instead of xaml markup?

**Comment by sotanakamura** [comment]
@saint4eva Yes. Writing UIs in Python is useful to set a event handler and bind some data. Of cource, you can also write the UIs in XAML.

**Comment by sotanakamura** [comment]
## Update

Constructors are now simpler.

```py
from win32more.Microsoft.UI.Xaml import Window
from win32more.Microsoft.UI.Xaml.Controls import ColorPicker
from win32more.xaml import XamlApplication


class App(XamlApplication):
    def OnLaunched(self, args):
        win = Window()
        clr = ColorPicker()
        win.Content = clr
        win.Activate()


XamlApplication.Start(App)
```

**Comment by jhi2** [comment]
> Hey iirc there _is_ a python library for this. It's called [`pythonnet`](https://github.com/pythonnet/pythonnet), a python library for the .NET framework.

Actually, .NET is not the same as winUI3

**Comment by jhi2** [comment]
I would love to work with C developers to make a python  binding for WinUI3

**Comment by jhi2** [comment]
> Hey iirc there _is_ a python library for this. It's called [`pythonnet`](https://github.com/pythonnet/pythonnet), a python library for the .NET framework.

.NET is not winui3

**Comment by jhi2** [comment]
> ## Update
> Constructors are now simpler.
> 
> ```python
> from win32more.Microsoft.UI.Xaml import Window
> from win32more.Microsoft.UI.Xaml.Controls import ColorPicker
> from win32more.xaml import XamlApplication
> 
> 
> class App(XamlApplication):
>     def OnLaunched(self, args):
>         win = Window()
>         clr = ColorPicker()
>         win.Content = clr
>         win.Activate()
> 
> 
> XamlApplication.Start(App)
> ```

WOW THAT IS SOOOOOOOOO COOL

**Comment by jhi2** [comment]
I am working on a wINui3 guI  TOOLKIT for Python see jhi2/PyUI3

**Comment by johnzhou721** [comment]
> > 朋友们，我找到了这个，它可以在 PyQt 或 Pyside 中使用一些流畅的小部件。https://github.com/zhiyiYo/PyQt-Fluent-Widgets
> 
> But is's not real WinUI :(

Agreed. PyQt is semi-native in terms of it using stateless system APIs to draw widgets, but it composites them into their own framebuffer, so it lacks native *feel*, e.g. modal diaglogs.

Also: YES, PLEASE!!!!! This has been almost FIVE years (3 months short of), and there has been no official response.  What is going on here, Microsoft???  Thanks!!!!

**Comment by johnzhou721** [comment]
@weitzhandler No idea... but in my perhaps flawed opinion if we get a Python binding done quickly more people will use it which means there will be more incentive to develop WinUI and more bugs wuold be found and fixed.

Unofficial binding here by the way: https://github.com/ynkdir/py-win32more

**Comment by S-1-5-21-1726115125-1917** [comment]
现在ynkdir/win32more已经足够完善了:
```
#type:ignore
from win32more.Microsoft.UI.Xaml import Window,Application
from win32more.Microsoft.UI.Xaml.Markup import IComponentConnector
from win32more.appsdk.xaml import XamlApplication,XamlLoader,ComClass
from win32more.Windows.Foundation import Uri
from typing_extensions import override
from pathlib import Path

class MainWindow(ComClass,Window,IComponentConnector):
    def __init__(self):
        super().__init__(own=True)
        self.InitializeComponent()
    
    def InitializeComponent(self):
        XamlPath=Path(__file__).with_name("MainWindow.xaml").absolute().as_posix()
        ResourceLocator=Uri(f"ms-appx:///{XamlPath}")
        Application.LoadComponent(self,ResourceLocator)
    
    def Connect(self,connectionId,target):
        pass #暂无

class App(XamlApplication):
    @override
    def OnLaunched(self,args):
        self.MainWindow=MainWindow()
        self.MainWindow.Activate()

XamlApplication.Start(App)
```

**Comment by mominshaikhdevs** [comment]
* https://github.com/microsoft/microsoft-ui-xaml/issues/2488#issuecomment-3341830565.

**Comment by DeZepTup** [comment]
Just curious, what is the current status?

**Comment by johnzhou721** [comment]
> Just curious, what is the current status?

The current status is that there's sort-of reliable external tools like py-win32more and pywinrt that you can use to build WinUI3 apps without Python, however they're unofficial.

Disclaimer I'm not related to MS.
